### PR TITLE
[xds] tests to wait for asynchronous metrics reports

### PR DESCRIPTION
**Pull request by yousukseung**
Reports to the metrics reporter may happen asynchronously with work serializer dispatch.

### CODE REVIEW

File: test/core/xds/xds_client_test.cc [hunk 1]
```diff
@@ -631,11 +656,25 @@ class XdsClientTest : public ::testing::Test {
       MutexLock lock(&mu_);
       return resource_updates_valid_;
     }
+    const ResourceUpdateMap* resource_updates_valid_ptr() const {
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:659]: Exposing all of this internal state is ugly and shouldn't be necessary.  Instead, I suggest something like this:

```
// Returns true if the matcher returns true before the timeout.
// Runs the predicate once as soon as it is called and then again
// every time the metrics reporter sees an update.
bool WaitForMetricsReporterData(
    ::testing::Matcher<ResourceUpdateMap> resource_updates_valid_matcher,
    ::testing::Matcher<ResourceUpdateMap> resource_updates_invalid_matcher,
    ::testing::Matcher<ServerFailureMap> server_failures_matcher,
    absl::Duration timeout = absl::Seconds(3),
    SourceLocation location = SourceLocation()) {
  const absl::Time deadline = absl::Now() + (timeout * grpc_test_slowdown_factor());
  MutexLock lock(&mu_);
  while (true) {
    if (::testing::Matches(resource_updates_valid_matcher)(resource_updates_valid_) &&
        ::testing::Matches(resource_updates_invalid_matcher)(resource_updates_invalid_) &&
        ::testing::Matches(server_failures_matcher)(server_failures_)) {
      return true;
    }
    if (cond_.WaitWithDeadline(&mu_, deadline)) break;
  }
  EXPECT_THAT(resource_updates_valid_, resource_updates_valid_matcher)
      << location.file() << ":" << location.line();
  EXPECT_THAT(resource_updates_invalid_, resource_updates_invalid_matcher)
      << location.file() << ":" << location.line();
  EXPECT_THAT(server_failures_, server_failures_matcher)
      << location.file() << ":" << location.line();
  return false;
}
```

Then the tests can invoke it using something like this:

```
EXPECT_TRUE(metrics_reporter_->WaitForMetricsReporterData(
    ::testing::ElementsAre(::testing::Pair(
        ::testing::Pair(kDefaultXdsServerUrl,
                        XdsFooResourceType::Get()->type_url()),
        1)),
    ::testing::_, ::testing::_));
```

File: test/core/xds/xds_client_test.cc [hunk 2]
```diff
@@ -81,6 +83,29 @@ namespace {
...
 constexpr absl::string_view kDefaultXdsServerUrl = "default_xds_server";
 
+// Attempts to match `inner_matcher` each time `ctx.condvar` is signaled or
+// `ctx.timeout` expires. Does not fail until the timeout expires.
+MATCHER_P2(PointeeMatchesWithTimeout, inner_matcher, ctx,
+           absl::StrCat("pointee matches with timeout ", ctx.timeout)) {
+  MutexLock lock(&ctx.mu);
+  while (true) {
+    if (::testing::Matches(::testing::Pointee(inner_matcher))(arg)) return true;
+    if (ctx.condvar.WaitWithTimeout(&ctx.mu, ctx.timeout)) break;
```
[the remark this answers is not in the document]
Comment (yousukseung) [reply test/core/xds/xds_client_test.cc]: Done.

File: test/core/xds/xds_client_test.cc [hunk 3]
```diff
@@ -637,6 +637,37 @@ class XdsClientTest : public ::testing::Test {
     }
     const ServerFailureMap& server_failures() const { return server_failures_; }
 
+    // Returns true if the matcher returns true before the timeout.
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:640]: s/matcher/matchers/

File: test/core/xds/xds_client_test.cc [hunk 4]
```diff
@@ -637,6 +637,37 @@ class XdsClientTest : public ::testing::Test {
...
+    // Runs the predicate once as soon as it is called and then again
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:641]: s/predicate/matchers/

File: test/core/xds/xds_client_test.cc [hunk 5]
```diff
@@ -930,19 +964,12 @@ TEST_F(XdsClientTest, BasicWatch) {
...
   // Check metric data.
-  EXPECT_THAT(metrics_reporter_->resource_updates_valid(),
-              ::testing::ElementsAre(::testing::Pair(
-                  ::testing::Pair(kDefaultXdsServerUrl,
-                                  XdsFooResourceType::Get()->type_url()),
-                  1)));
-  EXPECT_THAT(metrics_reporter_->resource_updates_invalid(),
-              ::testing::ElementsAre());
-  EXPECT_THAT(
-      GetResourceCounts(),
```
Comment (markdroth) [inline test/core/xds/xds_client_test.cc:941]: Looks like you accidentally removed this check.

**Review by markdroth — COMMENTED** [review]
Just one noteworthy issue remaining.

File: test/core/xds/xds_client_test.cc (hunk 3 above)
Comment (yousukseung) [reply test/core/xds/xds_client_test.cc]: Done

File: test/core/xds/xds_client_test.cc (hunk 4 above)
Comment (yousukseung) [reply test/core/xds/xds_client_test.cc]: Done

File: test/core/xds/xds_client_test.cc (hunk 5 above)
Comment (yousukseung) [reply test/core/xds/xds_client_test.cc:941]: Done

**Review by markdroth — APPROVED** [review]
Looks great!

**Pull request closed without merging**